packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -432,34 +432,66 @@ export class FluidDataStoreRuntime
 		return context.getChannel();
 	}
 
-	public createChannel(id: string = uuid(), type: string): IChannel {
+	public addChannel(channel: IChannel): void {
+		const id = channel.id;
 		if (id.includes("/")) {
 			throw new UsageError(`Id cannot contain slashes: ${id}`);
 		}
 
 		this.verifyNotClosed();
 
+		assert(!this.contexts.has(id), "addChannel() with existing ID");
+
+		const type = channel.attributes.type;
+		const factory = this.sharedObjectRegistry.get(channel.attributes.type);
+		if (factory === undefined) {
+			throw new Error(`Channel Factory ${type} not registered`);
+		}
```

Comment:
This bit seems strange to me - if we've already created the channel, why are we verifying the factory registration here?

Reply:
The user is adding a channel by creating it first. So, we are saying that, that channel should have been present in the registry and is not some random type of channel.

Reply:
Understood that it's important to ensure we don't try to attach a bogus channel, but my point is more that if we fail here then it means we've allowed the customer to create a bogus channel (and possibly populate it with data which will then be lost).  Having the check here is halfway through the bad thing happening, rather than at the start.

In creating the channel, we pass the runtime to `IChannelFactory.create`.  Maybe we could fail the factory creation at that point (i.e. say "You tried to create an instance using a runtime that doesn't include this factory in its registry" or something), well before we permit the attempt to add it here?

It would be helpful to see this in an example usage that looks more like a normal usage scenario, rather than just automated tests - real users won't be calling `createRuntime()` or casting JS objects to `IChannel`, so it's a bit challenging to understand if there are other opportunities to guard against mistakes.

Reply:
1.) DataStoreRuntime does not expose the registry on its surface and that is internal to the runtime. Also, the ChannelFactory does not have these registries which would fail this creation and they should not the registry, as they are just factories to create the particular channels.
2.) The channel interfaces for example, ISharedDirectory which is the return type of factory.create() or SharedObjects all extends IChannels , so there is no need of casting anything to anything.
3.) Also, this is not runtime behavior in the sense that, apps using this path would know that they cannot attach any channel to the runtime, and it needs to be in the registry, so they would know beforehand and would proceed accordingly.

Reply:
Added the comment on the api for the same.